lib/puppet/node.rb

```diff
@@ -108,12 +108,29 @@ def fact_merge
   # Merge any random parameters into our parameter list.
   def merge(params)
     params.each do |name, value|
-      @parameters[name] = value unless @parameters.include?(name)
+      if @parameters.include?(name)
+        Puppet::Util::Warnings.warnonce("The node paramters #{name} for node #{@parameters["hostname"]} was already set to #{@parameters[name]}. It could not be set to #{value}")
+      else
+        @parameters[name] = value
+      end
     end
 
     @parameters["environment"] ||= self.environment.name.to_s
   end
 
+  def add_server_facts(server_facts)
+   # Append the current environment to the list of server facts
+    complete_server_facts = server_facts.merge({ "environment" => self.environment.name.to_s})
+
+   # Set the top scope variable $server_facts if :trusted_server_facts is true
+   if Puppet[:trusted_server_facts]
+     @topscope.set_trusted(complete_server_facts)
+   end
+
+    # Merge the server facts into the parameters for the node
+    merge(server_facts)
```

Comment:
I was hesitant to do this because `merge` also adds the environment to the parameters and is called from places other than `add_server_facts`. So if I remove the addition of environment from `merge` it might not be added when it should be, but obviously when I leave it then it will be added twice. Alternately I could check in merge if it's already there before adding it. Not sure what's best.